File: src/main/java/io/github/jsnimda/common/config/IConfigElementResettableMultiple.java
package io.github.jsnimda.common.config;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

public interface IConfigElementResettableMultiple extends IConfigElementResettable {

  default Map<String, ? extends IConfigOption> getConfigOptionsMapFromConfigOptionsCollection() {
    Map<String, IConfigOption> map = new HashMap<>();
    for (IConfigOption configOption : getConfigOptions()) {
      map.put(configOption.getKey(), configOption);
    }
    return map;
  }

  Map<String, ? extends IConfigOption> getConfigOptionsMap();

  default Collection<? extends IConfigOption> getConfigOptionsCollectionFromConfigOptionsMap() {
    return getConfigOptionsMap().values();
  };

  Collection<? extends IConfigOption> getConfigOptions();

  @Override
  default JsonElement toJsonElement() {
    JsonObject obj = new JsonObject();
    for (IConfigOption configOption : getConfigOptions()) {
      if (configOption.isModified()) {
        obj.add(configOption.getKey(), configOption.toJsonElement());
      }
    }
    return obj;
  }

  @Override
  default void fromJsonElement(JsonElement element) {
    // reset to default first
    resetToDefault();
    if (element.isJsonObject()) {
      for (Entry<String, JsonElement> entry : element.getAsJsonObject().entrySet()) {
        Map<String, ? extends IConfigOption> configOptionsMap = getConfigOptionsMap();
        if (configOptionsMap.containsKey(entry.getKey())) {
          configOptionsMap.get(entry.getKey()).fromJsonElement(entry.getValue());
        } else {
          // TODO fail handle
        }
      }
    } else {
      // TODO fail handle
    }
  }
  
  @Override
  default boolean isModified() {
    return getConfigOptions().stream().anyMatch(x -> x.isModified());
  }

  @Override
  default void resetToDefault() {
    getConfigOptions().forEach(x -> x.resetToDefault());
  }

}

File: src/main/java/io/github/jsnimda/common/config/IConfigElementResettableMultiple.kt
package io.github.jsnimda.common.config

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import io.github.jsnimda.common.Log

interface IConfigElementResettableMultiple : IConfigElementResettable {

  // sub class should impl one of the getConfigOptionsMap() or getConfigOptions()

  fun getConfigOptionsMapFromList(): Map<String, IConfigOption> = getConfigOptionsList().associateBy { it.key }

  fun getConfigOptionsMap(): Map<String, IConfigOption>

  fun getConfigOptionsListFromMap(): List<IConfigOption> = getConfigOptionsMap().values.toList()

  fun getConfigOptionsList(): List<IConfigOption>

  override fun toJsonElement(): JsonElement = JsonObject().apply {
    getConfigOptionsList().forEach {
      if (it.isModified) this.add(it.key, it.toJsonElement())
    }
  }

  override fun fromJsonElement(element: JsonElement) { // reset to default first
    resetToDefault()
    try {
      val configOptionsMap = getConfigOptionsMap()
      element.asJsonObject.entrySet().forEach { (key, value) ->
        configOptionsMap[key].run {
          if (this == null) {
            Log.warn("[invprofiles.common] Unknown config key '$key' with value '$value'")
          } else {
            fromJsonElement(value!!)
          }
        }
      }
    } catch (e: Exception) {
      Log.warn("[invprofiles.common] Failed to set config value as [JsonObject] from the JSON element '$element'")
    }
  }

  override val isModified get() = getConfigOptionsList().any { it.isModified }

  override fun resetToDefault(): Unit = getConfigOptionsList().forEach { it.resetToDefault() }
}